File: cpp/367_Valid_Perfect_Square.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
	bool isPerfectSquare(int x) {
		if (x < 2) return true;

		long long l = 1, r = x;

		while (l <= r) {
			long long mid = l + (r - l) / 2;
			long long sq = mid * mid;

			if (sq == x) return true;
			else if (sq < x) l = mid + 1;
			else r = mid - 1;
		}
		return false;
	}
};

File: cpp/367_Valid_Perfect_Square.py
class Solution:
    def isPerfectSquare(self, x: int) -> bool:
        if x < 2:
            return True

        l, r = 1, x

        while l <= r:
            mid = l + (r - l) // 2
            sq = mid * mid

            if sq == x:
                return True
            elif sq < x:
                l = mid + 1
            else:
                r = mid - 1

        return False

